### ISSUE TITLE
[tune] Not returning correct best score

### ISSUE THREAD

**Issue by himat** [ray-project/ray#5174]
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 16.04
- **Ray installed from (source or binary)**: binary 
- **Ray version**: 0.7.2
- **Python version**: 3.6.5

### Describe the problem

It doesn't seem like tune is properly storing the best run?
When I use the experiment analysis tools to get the best trial from a given metric, here `mean_f1`, I am not getting the highest value from that run that I see in tb.

Does tune store the best `mean_f1` that was reported for a given trial, or is it keeping an average of all the `mean_f` values for that trial, or is it only keeping the `mean_f` from the last iteration of the trial, or something else? 

Here, it says the best trial has a mean_f1 of 0.5472222222222223.
```
>>> ana = ExperimentAnalysis(experiment_path=...)
>>> ana.get_best_info("mean_f1")
{'trainable_name': 'tune_trainable', 'local_dir': './experiments-tune/proj0_data_geosplit_7layers_newdata_downsamp_smaller_max_lr', 'experiment_tag': '152_lr=0.0014916,momentum=0.23592,sched_decay=0.48218,sched_step_size=22,weight_decay=0.056616', 'upload_dir': '', 'loggers': '80049502000000000000004e2e', 'sync_function': '80049502000000000000004e2e', 'verbose': True, 'max_failures': 3, 'last_update_time': 1562817103.585816, 'checkpoint_freq': 0, 'checkpoint_at_end': False, 'history': [], 'keep_checkpoints_num': None, '_cmp_greater': True, 'best_checkpoint_attr_value': -inf, 'checkpoint_score_attr': 'training_iteration', '_checkpoint': '80049553000000000000008c0e7261792e74756e652e747269616c948c0a436865636b706f696e749493942981947d94288c0773746f72616765948c046469736b948c0576616c7565944e8c0b6c6173745f726573756c74947d9475622e', 'export_formats': [], 'status': 'TERMINATED', 'logdir': './experiments-tune/proj0_data_geosplit_7layers_newdata_downsamp_smaller_max_lr/tune_trainable_152_lr=0.0014916,momentum=0.23592,sched_decay=0.48218,sched_step_size=22,weight_decay=0.056616_2019-07-11_03-40-23ptew416n', 'runner': None, 'result_logger': None, 'last_debug': 1562817103.5858095, 'trial_id': 'e89d14cc', 'error_file': None, 'num_failures': 0, 'custom_trial_name': None, 'results': '80049502000000000000004e2e', 'best_result': '80049502000000000000004e2e', 'param_config': '80049502000000000000004e2e', 'extra_arg': '80049502000000000000004e2e', '_nonjson_fields': ['_checkpoint', 'loggers', 'sync_function', 'results', 'best_result', 'param_config', 'extra_arg'], '__logger_started__': False, 'lr': 0.0014916372822960891, 'momentum': 0.2359183713580206, 'weight_decay': 0.056615924140151375, 'sched_step_size': 22, 'sched_decay': 0.4821802538173813, 'mean_loss': 0.5535685022672018, 'mean_f1': 0.5472222222222223, 'time_this_iter_s': 7.203242778778076, 'done': True, 'timesteps_total': None, 'episodes_total': None, 'training_iteration': 100, 'neg_mean_loss': -0.5535685022672018, 'experiment_id': '4c3081e8824f45e3871c808536daff6a', 'date': '2019-07-11_03-51-43', 'timestamp': 1562817103, 'time_total_s': 677.6749112606049, 'pid': 31902, 'hostname': 'ip-172-31-20-119', 'node_ip': '172.31.20.119', 'time_since_restore': 677.6749112606049, 'timesteps_since_restore': 0, 'iterations_since_restore': 100, 'config:lr': 0.0014916372822960891, 'config:momentum': 0.2359183713580206, 'config:weight_decay': 0.056615924140151375, 'config:sched_step_size': 22, 'config:sched_decay': 0.4821802538173813, 'resources:cpu': 2, 'resources:gpu': 1, 'resources:extra_cpu': 0, 'resources:extra_gpu': 0, 'stopping_criterion:mean_f1': 0.99}

```

But from tb, I saw multiple higher values for that same trial number 152. Here is one at 0.563:
![image](https://user-images.githubusercontent.com/1156974/61073514-9a1ce780-a3e3-11e9-895f-e29f71e8fec2.png)
And the last value I saw on tb is 0.549, which is still higher than the tune reported one.

The tune code I used to run the experiment:
```
scheduler = schedulers.AsyncHyperBandScheduler(
    time_attr='training_iteration',
    reward_attr='neg_mean_loss',
    max_t=100,
    grace_period=10,
    reduction_factor=3,
    brackets=3
)

all_trials = tune.run(
    tune_trainable,
    name=exp_name,
    local_dir=tune_exp_dir,
    scheduler=scheduler,
    stop={"mean_f1": 0.99},
    resources_per_trial={"cpu": 2, "gpu": 1},
    num_samples=360, 
    config={
        "lr": tune.sample_from(lambda spec: np.random.uniform(1e-12, 0.4)),
        "momentum": tune.sample_from(lambda spec: np.random.uniform(0.0, 0.4)),
        "weight_decay": tune.sample_from(lambda spec: np.random.uniform(0.0, 1.0)),
        "sched_step_size": tune.sample_from(lambda spec: np.random.randint(1, 100)),
        "sched_decay": tune.sample_from(lambda spec: np.random.uniform(0.0, 1.0))
        }
)
```

And in my training loop, after running on validation data, I use this line to report the `mean_f1` as just the f1 computed on all the validation data predictions.
```
self.tune_reporter(mean_loss=val_loss, mean_f1=val_f1)
```

**Comment by richardliaw** [comment]
Hm, do you mind posting a zip of your log files?

**Comment by himat** [comment]
Also, I checked the experiment_state json, and it looks like that best value that's being reported above is actually in the "last_result" portion of the config for trial 152. But that still seems weird since that's not the last value I see in tb.

![image](https://user-images.githubusercontent.com/1156974/61085652-c72ac380-a3fe-11e9-8a30-10a6695cb546.png)

Here's the experiment log directory zip: https://www.dropbox.com/s/0eiy83crz4p7rh5/tune_exp_dir.tar.gz?dl=0
- (just a note that doesn't affect anything, but I only wrapped tune around my already existing pytorch project instead of inheriting from your class, so that's why there's an experiments/ folder within each trial where I store my own tb logs and configs instead of using Tune's built-in checkpointing and things.)

Thanks Richard!
